Byfield gpu

rohanchanani opened the pull request:
Implements the by field operator for GPUs. To build, configure cmake with -DREALM_USE_CUDART_HIJACK=OFF (necessary) and -DCMAKE_CUDA_ARCHITECTURES={your arch} (to speed up build). To test, run ./tests/deppart basic -ll:gpu 1
@rohany

elliottslaughter committed:
Bump version.

muraj committed:
realm: fix and add capi tests

muraj committed:
Merge branch 'cperry/capi-ci' into 'master'

realm: fix and add capi tests

See merge request StanfordLegion/legion!1812

elliottslaughter committed:
test: Match the allocation function used in test.

lightsighter committed:
legion: add extra reference to avoid overzealous assertions in debug mode

elliottslaughter committed:
Merge branch 'fix-test-barrier-arrivals' into 'master'

[P0] test: Match the allocation function used in test

See merge request StanfordLegion/legion!1817

elliottslaughter committed:
Merge branch 'fix-test-barrier-arrivals' into 'master'

[P0] test: Match the allocation function used in test

See merge request StanfordLegion/legion!1817

(cherry picked from commit a101776902a4124e34bf3e371ac2828c318dafac)

f753d5d5 test: Match the allocation function used in test.

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

eddy16112 committed:
update realm release notes for 25.06

elliottslaughter committed:
Merge branch 'rc-realm-notes' into 'rc'

update realm release notes for 25.06

See merge request StanfordLegion/legion!1818

elliottslaughter committed:
Update CHANGES.txt.

elliottslaughter committed:
Merge branch 'stable'

rainmakereuab committed:
Apply 1 suggestion(s) to 1 file(s)

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

muraj committed:
realm: update gitlab-ci to build realm externally and fix bootstrap_mpi in new build system

muraj committed:
Merge branch 'cperry/ci-changes' into 'master'

realm: update gitlab-ci to build realm externally and fix bootstrap_mpi in new build system

See merge request StanfordLegion/legion!1820

lightsighter committed:
legion/mapping: add a case for read-discard to mapping utilities

lightsighter committed:
legion/mapping: fix dumb compiler warning

elliottslaughter committed:
regent: Build Legion with CMake by default. Remove options to build Terra without CMake.

elliottslaughter committed:
test: Re-enable attach tests on macOS.

eddy16112 committed:
Refactor DynamicTable such that we can pass a RuntimeImpl to the elements of DynamicTableNode

eddy16112 committed:
Merge branch 'refactor-dynamictable' into 'master'

Refactor DynamicTable such that we can pass a RuntimeImpl to the elements of DynamicTableNode

See merge request StanfordLegion/legion!1814

elliottslaughter committed:
Merge branch 'test-attach-file-mini' into 'master'

test: Re-enable attach tests on macOS

See merge request StanfordLegion/legion!1826

elliottslaughter committed:
regent: Tweak build for CI to default to Make when USE_CMAKE is unset.

elliottslaughter committed:
regent: Print commands being executed by travis.py.

elliottslaughter committed:
ci: Refactor flush calls.

elliottslaughter committed:
regent: Fix CMake flag.

muraj committed:
Merge branch 'cperry/legion-cmake-rework' into 'master'

legion: Modify legion cmake to pick up the new realm build system

See merge request StanfordLegion/legion!1644

muraj committed:
legion: Modify legion cmake to pick up the new realm build system

elliottslaughter committed:
Merge branch 'regent-cmake-default' into 'master'

regent: Build Legion with CMake by default

See merge request StanfordLegion/legion!1825

elliottslaughter committed:
ci: Run MSVC jobs on AVX only.

elliottslaughter committed:
Merge branch 'ci-mscv-avx' into 'master'

ci: Run MSVC jobs on AVX only

See merge request StanfordLegion/legion!1828

Commit:
test htr ci

Commit:
disable soleil external test

Commit:
soleil test

seemamirch committed:
Merge branch 'sm/fix_htr_ci' into 'master'

Enable HTR in CI tests

See merge request StanfordLegion/legion!1830

muraj committed:
realm: initiate final move of source files to realm subdir

muraj committed:
Merge branch 'cperry/realm-src-move' into 'master'

realm: initiate final move of source files to realm subdir

See merge request StanfordLegion/legion!1827

Commit:
add htr+soleil tests back

elliottslaughter committed:
realm: Restore functionality of hipGetTaskStream without hijack.

seemamirch committed:
Merge branch 'fix_htr_soleil_ci' into 'master'

Enable htr+soleil tests in Legion CI

See merge request StanfordLegion/legion!1832

muraj committed:
realm: Fix up static builds part 1

muraj committed:
Merge branch 'cperry/fix-static-deps' into 'master'

realm: Fix up static builds part 1

See merge request StanfordLegion/legion!1833

eddy16112 committed:
Realm: fix the proc-mem affinity for GPUs without discrete FB memory

eddy16112 committed:
Merge branch 'fix-best-affinity-dgx' into 'master'

Realm: fix the proc-mem affinity for GPUs without discrete FB memory

See merge request StanfordLegion/legion!1836

elliottslaughter committed:
build: Remove explicit configuration of -fPIC, which is default anyway.

eddy16112 committed:
Realm: use realm status for the module config

eddy16112 committed:
Merge branch 'fix-module-config-warning' into 'master'

Realm: use realm status for the module config

See merge request StanfordLegion/legion!1839

elliottslaughter committed:
Merge branch 'embed-gasnet-fix-shared' into 'master'

build: Remove explicit configuration of -fPIC for GASNet, which is default anyway

See merge request StanfordLegion/legion!1840

muraj committed:
Merge branch 'cperry/disable-auto-features' into 'master'

realm: default disable dependencies that would be downloaded if not found.

See merge request StanfordLegion/legion!1838

muraj committed:
realm: default disable dependencies that would be downloaded if not found.

muraj committed:
realm: Update LegionConfig.cmake to look for Realm and make sure Realm looks for Threads

muraj committed:
Merge branch 'cperry/fix-examples-build' into 'master'

realm: Update LegionConfig.cmake to look for Realm and make sure Realm looks for Threads

See merge request StanfordLegion/legion!1841

muraj committed:
realm: Fix up HIP builds

muraj committed:
Merge branch 'cperry/fix-hip' into 'master'

realm: Fix up HIP builds

See merge request StanfordLegion/legion!1843

elliottslaughter committed:
ci: Add module load for Frontier Python.

elliottslaughter committed:
ci: Switch Frontier to new CI root.

elliottslaughter committed:
ci: Add proxy settings for Frontier.

elliottslaughter committed:
Merge branch 'hip-fix-task-stream' into 'master'

[P0] realm: Restore functionality of hipGetTaskStream without hijack

See merge request StanfordLegion/legion!1831

eddy16112 committed:
Merge branch 'capi-machine-addressapce' into 'master'

Realm: address space C API

See merge request StanfordLegion/legion!1846

eddy16112 committed:
Realm: add address space C API

eddy16112 committed:
Realm: add C API for runtime library version

eddy16112 committed:
Merge branch 'capi-runtime-version' into 'master'

Realm: add runtime version for C API

See merge request StanfordLegion/legion!1845

elliottslaughter committed:
ci: Make Frontier CI files group-writable so they can be deleted.

elliottslaughter committed:
ci: Set umask on Frontier CI.

elliottslaughter committed:
ci: Ensure Frontier CI files are group writable.

elliottslaughter committed:
ci: Always chmod files at end of test.

elliottslaughter committed:
ci: Run chmod even if test times out.

muraj committed:
legion: Add gasnet installation and modify config to use find_package

elliottslaughter committed:
ci: Specify HIP architecture for Frontier and disable Make tests for now.

muraj committed:
Merge branch 'cperry/embed-gasnet-install' into 'master'

legion: Add gasnet installation and modify config to use find_package

See merge request StanfordLegion/legion!1851

eddy16112 committed:
realm: C API for getting local address space

eddy16112 committed:
Merge branch 'capi-local-address-space' into 'master'

realm: C API for getting local address space

See merge request StanfordLegion/legion!1850

muraj committed:
realm: fix up version handling in the case that git fails to fetch tags and...

muraj committed:
Merge branch 'cperry/realm-version-fix' into 'master'

realm: fix up version handling in the case that git fails to fetch tags and...

See merge request StanfordLegion/legion!1852

Jacobfaib committed:
Fix NVTX target selection

Jacobfaib committed:
Merge branch 'jacobf/2025-07-25/realm-fix-nvtx-targets' into 'master'

Fix NVTX target selection

See merge request StanfordLegion/legion!1849

Jacobfaib committed:
Include nvtx3 headers if linking against nvtx3, not nvToolsExt

muraj committed:
realm: fix nvtx so it's not enabled if nvtx wasn't actually found

lightsighter committed:
Merge branch 'master' into asynceffects

lightsighter committed:
Merge branch 'master' into profasynceffect

lightsighter committed:
legion: always deactivate point tasks directly from slice tasks

muraj committed:
Merge branch 'cperry/fix-nvtx' into 'master'

realm: fix nvtx so it's not enabled if nvtx wasn't actually found

See merge request StanfordLegion/legion!1856

muraj committed:
realm: limit ccache to be just from master, reducing cache size needed to just one branch

Jacobfaib committed:
Merge branch 'jacobf/2025-07-28/fix-nvtx-target-2' into 'master'

Include nvtx3 headers if linking against nvtx3, not nvToolsExt

See merge request StanfordLegion/legion!1854

muraj committed:
realm: only set cmake_cuda_compiler if not already set

muraj committed:
realm: make rt library optional, and instead check for symbol support

muraj committed:
realm: add WAR for module libraries and VERSION keyword

muraj committed:
Merge branch 'cperry/osx-fix' into 'master'

realm: add WAR for module libraries and VERSION keyword

See merge request StanfordLegion/legion!1860

rainmakereuab committed:
Merge branch 'fixslicedeactive' into 'master'

Fix Slice Deactivate

See merge request StanfordLegion/legion!1859

lightsighter committed:
legion: tighten layout constraints for alignment for attach launcher helper methods

lightsighter committed:
legion: correctly identify when the ready queue state of a mapper has changed during a call to select_tasks_to_map

eddy16112 committed:
Realm: add C API for region instance

eddy16112 committed:
Merge branch 'capi-inst' into 'master'

Realm: add C API for instance

See merge request StanfordLegion/legion!1785

muraj committed:
realm: One more fix for NVTX - set the default based on REALM_ENABLE_CUDA

ebendler committed:
Add new header "realm.hpp" that sits on top of the C API

ebendler committed:
Merge branch 'realm_small_hpp' into 'master'

Add new header "realm.hpp" that sits on top of the C API

See merge request StanfordLegion/legion!1824

lightsighter committed:
Merge branch 'master' into fixmapperevent

rainmakereuab committed:
Merge branch 'fixmapperevent' into 'master'

Fix Mapper Event Scheduling

See merge request StanfordLegion/legion!1864

muraj committed:
Merge branch 'cperry/ccache-reduce' into 'master'

realm: limit ccache to be just from master, reducing cache size needed to just one branch

See merge request StanfordLegion/legion!1861

rainmakereuab committed:
Merge branch 'profasynceffect' into 'master'

Profile Asynchronous Effects

See merge request StanfordLegion/legion!1796

muraj committed:
realm: Add enable_language(HIP) support to realm build system

muraj committed:
Merge branch 'cperry/hip-fix' into 'master'

realm: Add enable_language(HIP) support to realm build system

See merge request StanfordLegion/legion!1855

muraj committed:
Merge branch 'cperry/rt-optional' into 'master'

realm: make rt library optional, and instead check for symbol support

See merge request StanfordLegion/legion!1863

elliottslaughter committed:
legion: Fix trie test.

elliottslaughter committed:
ci: Avoid copying files around for Frontier CI.

Jacobfaib committed:
Properly disable the altstack on macOS as well

Jacobfaib committed:
Merge branch 'jacobf/2025-07-30/fix-altstack-macos' into 'master'

Properly disable the altstack on macOS as well

See merge request StanfordLegion/legion!1865

elliottslaughter committed:
Merge branch 'ci-frontier-updates'

elliottslaughter committed:
Merge branch 'fix-trie' into 'master'

legion: Fix trie test

See merge request StanfordLegion/legion!1868

lightsighter committed:
legion: make sure we check whether color spaces contain points before trying to linearize them

eddy16112 committed:
Realm: add faultware to event C API

eddy16112 committed:
Merge branch 'capi-event-faults' into 'master'

Realm: add fault to event C API

See merge request StanfordLegion/legion!1842

muraj committed:
Merge branch 'cperry/wrap-cmake-cuda' into 'master'

realm: only set cmake_cuda_compiler if not already set

See merge request StanfordLegion/legion!1862

muraj committed:
realm: Fix cuGetProcAddress versioning

muraj committed:
Merge branch 'cperry/cuda-loader-fix' into 'master'

realm: Fix cuGetProcAddress versioning

See merge request StanfordLegion/legion!1866

muraj committed:
Merge branch 'cperry/grand-format' into 'master'

The grand reformatting

See merge request StanfordLegion/legion!1869

muraj committed:
The grand reformatting

muraj committed:
realm: Ignore formatting commit in git blame

muraj committed:
realm: Fix up hip+gasnet builds

muraj committed:
Merge branch 'cperry/gasnet-hip-fix' into 'master'

realm: Fix up hip+gasnet builds

See merge request StanfordLegion/legion!1870

elliottslaughter committed:
ci: Move checkouts back to home directory.

muraj committed:
Update copyright and licenses across all files

muraj committed:
Merge branch 'cperry/reuse-compliance' into 'master'

Update copyright and licenses across all files

See merge request StanfordLegion/legion!1873

ebendler committed:
Update realm.hpp events to match realm_c.h changes. Add missing docstring for wait_on param on realm_region_instance_destroy.

muraj committed:
Change GASNet_ROOT_DIR->GASNet_ROOT everywhere.

lightsighter committed:
legion: prevent instances mapped by a task with an unbounded pool in the same memory from being used to satisfy deferred deletions/allocations during the execution of that task

eddy16112 committed:
realm: do not copy ProfilingRequestSet for C API

eddy16112 committed:
Merge branch 'capi-rm-profiling-cp' into 'master'

realm: do not copy ProfilingRequestSet for C API

See merge request StanfordLegion/legion!1872

muraj committed:
Merge branch 'cperry/gasnet_root' into 'master'

Change GASNet_ROOT_DIR->GASNet_ROOT everywhere.

See merge request StanfordLegion/legion!1879

eddy16112 committed:
Merge branch 'fix-memcpy-benchmark' into 'master'

realm: fix the remote copy for memcpy benchmark

See merge request StanfordLegion/legion!1745

eddy16112 committed:
realm: fix the remote copy for memcpy benchmark

muraj committed:
realm: Various fixes found in github CI

muraj committed:
Merge branch 'cperry/github-update' into 'master'

realm: Various fixes found in github CI

See merge request StanfordLegion/legion!1874

muraj committed:
Remove all of legion

muraj committed:
Move realm source to it's new home

muraj committed:
Add git attributes and VERSION file for source archive versioning

muraj committed:
realm: Fix minor typo with event tutorial

muraj committed:
realm: couple small fixes for windows compilation

muraj committed:
realm: fix up coverage workflow

muraj committed:
realm: fix up test result reporting

eddy16112 committed:
realm: add C API for Event::has_triggered

eddy16112 committed:
Merge branch 'capi-event-has_triggered' into 'master'

realm: add C API for Event::has_triggered

See merge request StanfordLegion/legion!1875

rainmakereuab committed:
Merge branch 'fixunboundpool' into 'master'

Fix Unbounded Pool Deferred Allocations/Deletions

See merge request StanfordLegion/legion!1880

lightsighter committed:
legion: small fix for reference source kinds

muraj committed:
realm: Fix up github workflows and enable coverage

muraj committed:
realm: Various fixes to test_c_external_inst.cc to compile on windows

muraj committed:
realm: fix small typo in github workflow

muraj committed:
realm: fix misedit of test_c_external_inst.cc

muraj committed:
realm: one last fix to the copyright checker

muraj committed:
realm: ignore error from removal of build log

muraj committed:
Fix build log update and add codecov badge

muraj committed:
Add the workflow file as a trigger for updating code coverage

muraj committed:
Add remote trigger of gitlab pipelines

ebendler committed:
Merge branch 'capi-doxy-param' into 'master'

Add missing docstring for wait_on param on realm_region_instance_destroy

See merge request StanfordLegion/legion!1877

Jacobfaib committed:
Fix -Wmissing-template-arg-list-after-template-kw

elliottslaughter committed:
ci: Shut off failing Frontier CI tests for now.

elliottslaughter committed:
ci: Add new ROCm versions to Frontier CI.

elliottslaughter committed:
ci: Customize jobs to use fewer nodes.

muraj committed:
realm: make gasnetex default in legion and modify some logic for realm and gasnet

Jacobfaib committed:
Merge branch 'jacobf/2025-08-04/fix-wrong-template-kw' into 'master'

Fix -Wmissing-template-arg-list-after-template-kw

See merge request StanfordLegion/legion!1881

muraj committed:
realm: Add capability to replace realm in ci

muraj committed:
Merge branch 'cperry/add-realm-remote' into 'master'

realm: Add capability to replace realm in ci

See merge request StanfordLegion/legion!1883

lightsighter committed:
legion: fix locking bug for pointwise dependence analysis

ebendler committed:
Enable compile_commands.json generation for Realm

ebendler committed:
Merge branch 'compile_commands_return' into 'master'

Enable compile_commands.json generation for Realm

See merge request StanfordLegion/legion!1890

lightsighter committed:
realm: merge master into asynceffects and resolve conflicts with reformatting

muraj committed:
cuda: Fixes for 13.0 (#240)

* cuGetProcAddress broke it's own rules for a couple of apis and kept
back old versions of apis, but left cuGetProcAddress to return the newer
ones, breaking source compatibility guarentees for 13.0
* For those apis that were held back, an explicit context or location
was added, so modify the calls to pass in the required arguments, being
compatible with older toolkits as well.

muraj committed:
Merge branch 'cperry/gasnetex-default' into 'master'

realm: make gasnetex default in legion and modify some logic for realm and gasnet

See merge request StanfordLegion/legion!1886

eddy16112 committed:
realm: add namespace to ExternalInstanceResource classes

eddy16112 committed:
Merge branch 'fix-cuda-external-memory' into 'master'

realm: add namespace to ExternalInstanceResource classes

See merge request StanfordLegion/legion!1895

rainmakereuab committed:
Merge branch 'asynceffects' into 'master'

Finish Event Precondition

See merge request StanfordLegion/legion!1770

ebendler committed:
Turn on the cmake setting to generate a compile_commands.json file du… (#243)

…ring the build. This is required for clangd, et al, to function.

ebendler committed:
Add cursor's settings folder to the gitignore file (#244)

Also adds the gh copilot thinking file.

lightsighter committed:
examples: add clarifying documentation that you cannot mix handshake synchronization modes

lightsighter committed:
legion: fix a case for waiting on futures outside of Legion tasks

eddy16112 committed:
realm: add query instance attributes for the C API

eddy16112 committed:
Merge branch 'capi-inst-attr' into 'master'

realm: add query instance attributes for the C API

See merge request StanfordLegion/legion!1884

cryos committed:
CMake: Only use a feature if it is enabled (#242)

Change the behavior to check the target was defined and the feature was
enabled before setting it to be used/built. This is useful when the
project is built as part of other projects that may have found a package
but do not necessarily want that feature to be built by this project.

eddy16112 committed:
realm: fix a bug in the machine model C API

eddy16112 committed:
Merge branch 'capi-fix-machinemodel' into 'master'

realm: fix a bug in the machine model C API

See merge request StanfordLegion/legion!1848

muraj committed:
Various CI fixes (#248)

* Add sanitizer tests for ASAN, TSAN, and UBSAN
  * Unfortunately we have a leak in the python module, so disable that.
  * Fix a couple of found leaks
* Add ucx and gasnet builds to the ci

elliottslaughter committed:
ci: Run all Frontier tests on login12.

eddy16112 committed:
realm: C API for has_affinity calls

eddy16112 committed:
Merge branch 'capi-has-affinity' into 'master'

realm: C API for has_affinity calls

See merge request StanfordLegion/legion!1853

lightsighter committed:
realm: merge master into sized alloc and resolve conflicts

lightsighter committed:
realm: fix merge errors for size based range allocator

lightsighter committed:
realm: more fixes for broken merge of size based range allocator

lightsighter committed:
realm: more fixes sized based memory allocator merge

eddy16112 committed:
Merge branch 'capi-refactor-external-instance' into 'master'

realm: refactor the C API for realm_external_resource_t such that it becomes a plain struct instead of a handle.

See merge request StanfordLegion/legion!1892

eddy16112 committed:
realm: refactor the C API for realm_external_resource_t such that it becomes a plain struct instead of a handle.

lightsighter committed:
realm: switch back to the basic range allocator for now, reenable the sized based range allocator later

rainmakereuab committed:
Merge branch 'sizedalloc' into 'master'

Size-Based Range Allocator

See merge request StanfordLegion/legion!1394

muraj committed:
Rewrite logic to remove NDEBUG and apply it to all builds (#250)

lightsighter committed:
legion: fixes for auto-trace detection

lightsighter committed:
legion: more fixes for auto-tracing

lightsighter committed:
legion: more fixes for auto tracing

lightsighter committed:
legion: small fix for pending pointwise dependences

lightsighter committed:
legion: initialize shadow indirections for copy mappings

lightsighter committed:
legion: fix for shadow indirection setup

muraj committed:
Add ability to override test args at runtime (#253)

* Fix up cuda compile args to compile SASS for everything up to the last
architecture, adding ptx for the latest (fixing an issue with r580 when
both sass and ptx is available for a removed architecture)
* Add a REALM_DEFAULT_ARGS environment and put the default test
arguments here, prepended to what is set in the test environment at
runtime. This allows us to set the default, but allow the test
environment to override it.
* This allows us to, by default, enable GPU tests with one gpu, but
override it in the test environment for more than one GPU

ebendler committed:
Refactor the CAPI integration tests to use the CHECK_REALM macro (#258)

ebendler committed:
Add the ASSERT_REALM macro to machine_config_test to make it more readable (#259)

rainmakereuab committed:
Merge branch 'fixshadow' into 'master'

Fix Shadow Indirection Copies

See merge request StanfordLegion/legion!1903

lightsighter committed:
Merge branch 'master' into fixautotrace

lightsighter committed:
test: add a test to confirm that auto tracing is working

lightsighter committed:
test: make info the default realm logging level for all tests

lightsighter committed:
test: modify the auto tracing test to use the existing logfile

lightsighter committed:
test: fix auto tracing test to handle node numbers in logfile names

lightsighter committed:
test: get the name of the auto tracing logger correct

lightsighter committed:
test: more fixes for auto tracing test with cmake

rainmakereuab committed:
Merge branch 'fixautotrace' into 'master'

Fix Auto Tracing

See merge request StanfordLegion/legion!1901

muraj committed:
Remove accidently added autotracing test

muraj committed:
Add missing copyright information in test

ebendler committed:
Refactor OperationTable::request_cancellation (#255)

This changes request_cancellation to return a bool stating success/fail
and also refactors the flow to make it readable.

muraj committed:
Update github actions to build cuda 13.0 (#268)

muraj committed:
Fix cmake prepend to include a space in REALM_DEFAULT_ARGS

muraj committed:
Fix machine_config_test to check return value

fixup! Add the ASSERT_REALM macro to machine_config_test to make it more readable (#259)

ebendler committed:
Refactor realm_event_wait to take a timer parameter (#256)

This eliminates the need for timed_wait as a separate function. Then update the tests to the new API definition

Commit:
initial commit

Commit:
preparing to push

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+    cudaStream_t stream;
+    CUDA_CHECK(cudaStreamCreate(&stream), stream);
+
+    Memory my_mem = field_data[0].inst.get_location();
```
Are all the instances gauranteed to be in the same memory? Do you check this anywhere?

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+
+    nvtx_range_push("cuda", "build parent and instance entries");
+
+    std::vector<size_t> inst_offsets(field_data.size() + 1);
```
Needs a comment about what this is computing

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+    SparsityMapEntry<N,T>* d_inst_entries = reinterpret_cast<SparsityMapEntry<N,T>*>(AffineAccessor<char,1>(inst_entries_instance, 0).base);
+
+    size_t inst_pos = 0;
+    for (size_t i = 0; i < field_data.size(); ++i) {
```
I can't tell what's happening here but it seems like there should be a way to create an allocation then turn it into a single cudaMemcpy call.

rohany replied:
Yeah, I’m saying that you could probably write it to populate one buffer in host memory and then issue a single memcpy to the device.

rohany replied:
I still think this could be 1 cudaMemcpy.

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+
+    span<SparsityMapEntry<N,T>> parent_entries;
+    std::vector<SparsityMapEntry<N,T>> parent_entries_vec;
+    RegionInstance parent_entries_instance;
```
These didn't need to be declared without initialization

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+      parent_entries = parent_space.sparsity.impl()->get_entries();
+    }
+    parent_entries_instance = this->realm_malloc(parent_entries.size() * sizeof(SparsityMapEntry<N,T>), my_mem);
+    d_parent_entries = reinterpret_cast<SparsityMapEntry<N,T>*>(AffineAccessor<char,1>(parent_entries_instance, 0).base);
```
You could use these https://github.com/StanfordLegion/realm/blob/main/src/realm/instance.h#L133 methods probably if the AffineAccessor is cumbersome.

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+    CUDA_CHECK(cudaMemcpyAsync(d_inst_offsets, inst_offsets.data(), (field_data.size() + 1) * sizeof(size_t), cudaMemcpyHostToDevice, stream), stream);
+
+    int flattened_threads = 256;
+    int grid_size = (parent_entries.size() * inst_size + flattened_threads - 1) / flattened_threads;
```
this intermediate computation could probably overflow an int32 for a large instance.

rohany replied:
`int` by default is not necessarily int64_t, so this could still overflow. Also, it is common to extract this logic into a macro or function since it will be used repeatedly.

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+    intersect_input_rects<N,T><<<grid_size, flattened_threads, 0, stream>>>(d_inst_entries, d_parent_entries, d_inst_offsets, nullptr, inst_size, parent_entries.size(), field_data.size(), d_inst_counters, nullptr);
+    KERNEL_CHECK(stream);
+
+    uint32_t h_inst_counters[field_data.size()+1];
```
Why is this randomly a dynamic stack allocation?

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+
+    uint32_t h_inst_counters[field_data.size()+1];
+    h_inst_counters[0] = 0; // prefix sum starts at 0
+    CUDA_CHECK(cudaMemcpyAsync(h_inst_counters+1, d_inst_counters, field_data.size() * sizeof(uint32_t), cudaMemcpyDeviceToHost, stream), stream);
```
Why do you need to copy this back to the host to scan?

rohanchanani replied:
It's usually tiny (the number of instances) and I need the last value on the host anyways.

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    size_t big_inst_counters[field_data.size()+1];
+    for (size_t i = 0; i < field_data.size()+1; ++i) {
```
Is this not a memcpy?

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+
+    CUDA_CHECK(cudaMemsetAsync(d_inst_counters, 0, (field_data.size()) * sizeof(uint32_t), stream), stream);
+
+    flattened_threads = 256;
```
why redefine?

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+    nvtx_range_pop();
+    nvtx_range_push("cuda", "build volume offsets");
+
+    RegionInstance prefix_rects_instance = this->realm_malloc((num_valid_rects + 1) * sizeof(size_t), my_mem);
```
Is there any way you can be reusing these instances?

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+    RegionInstance valid_rects_instance = this->realm_malloc(num_valid_rects * sizeof(Rect<N,T>), my_mem);
+    Rect<N,T>* d_valid_rects = reinterpret_cast<Rect<N,T>*>(AffineAccessor<char,1>(valid_rects_instance, 0).base);
+
+    CUDA_CHECK(cudaMemsetAsync(d_inst_counters, 0, (field_data.size()) * sizeof(uint32_t), stream), stream);
```
Does your kernel require this data to be zero initialized?

rohanchanani replied:
Yes, these are the counters for the count + emit pass. On the first pass, it counts how many rectangles are to determine how big to alloc. On the second pass, each thread reuses the same counters to figure out where to write. I'll comment.

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+
+    nvtx_range_push("cuda", "populate bitmasks");
+
+    int threads_per_block = 256;
```
random redefinition of stuff above?

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+#pragma once
+#include "realm/deppart/byfield.h"
```
I've left comments on this file with mostly stylistic things. I'm not yet going to read the other parts of the pipeline, because they share the same problem as here: there are not enough comments or suggestive variable names that describe the overall picture of what this function is actually computing, and how the pieces fit together. Right now, it's not possible to follow the code as someone who didn't write it. This is a pretty algorithmic piece of code -- you should be able to describe in comments (maybe with small examples or descriptions or textual diagrams) about what each phase of the computation is doing, and how that fits into the overall algorithm.

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+      h_accessors[i] = AffineAccessor<FT,N,T>(field_data[i].inst, field_data[i].field_offset);
+    }
+
+    RegionInstance accessors_instance = this->realm_malloc(field_data.size() * sizeof(AffineAccessor<FT,N,T>), my_mem);
```
this seems like an example of an instance that could probably live in zero copy memory instead of device memory to avoid a copy.

rohany replied:
Bumping this, could just be a zero copy vector of accessors.

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+
+  CUDA_CHECK(cudaStreamSynchronize(stream), stream);
+  cudaStreamDestroy(stream);
+  inst_entries_instance.destroy();
```
There are instances you deleted here but not in the exit of the fast path. Also, do you need to wait until the end of the function to delete these? You should be trying to free (or reuse) these instances as soon as the data is no longer needed.

rohany reviewed (changes requested):
Initial round of feedback

rohany commented:
Realm has Gtest infrastructure now -- I think you should add your individual tests as unit tests following the patterns here: https://github.com/StanfordLegion/realm/tree/main/tests/unit_tests.

rohanchanani commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+#pragma once
```
Should I just make them .h? Or .cuh?

rohany replied:
just .h? be consistent with the style in the repo

rohanchanani commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,261 @@
+void GPUByFieldMicroOp<N,T,FT>::gpu_populate_bitmasks()
+{
+
+    nvtx_range_push("cuda", "byfield_gpu_populate_bitmasks");
```
Should I reduce the granularity of my nvtx ranges to avoid vomiting these guards all over the code? i.e. only push and pop at the start and end of each populate function?

rohany replied:
Having more detailed ranges is fine, if it helps you. To clean up the code you could probably write a class (add it to the nvtx file) that handles the ranges with RAII so it is simpler to look at. Something like the official nvtx api: https://nvidia.github.io/NVTX/. We can't actually use it directly for packaging reasons, but you could add a utility like this.

rohanchanani committed:
comments

rohanchanani commented on src/realm/deppart/partitions.cc:
```diff
@@ -71,7 +71,7 @@ namespace Realm {
 				      IndexSpace<N,T> *results,
 				      size_t first_result, size_t last_result,
-				      const std::vector<SparsityMapEntry<N,T> >& entries)
+				      const span<SparsityMapEntry<N,T> >& entries)
```
It errors when I do std::span, and elsewhere in realm is just span<> as well - it looks like it's coming from realm/utils.h with a comment that says:
// TODO: actually use C++20 version if available

rohanchanani commented on src/realm/sparsity.h:
```diff
@@ -246,8 +246,12 @@ namespace Realm {
+    std::vector<SparsityMapEntry<N,T> > entries;
+    span<SparsityMapEntry<N,T> > span_entries;
+    std::vector<Rect<N,T> > approx_rects;
+    span<Rect<N,T> > span_approx_rects;
```
Is it better to store instance + size and then a boolean for whether to use instance/vectors than to store the span?

rohanchanani replied:
Also, if I construct the span on the fly rather than storing it, should I change the function signature to not return by reference?

rohanchanani replied:
Yeah, its size is just a pointer and a size_t. Also, to pass it by reference I'd need to keep it as a member variable.

rohany replied:
There’s no need to return a reference then. References are either for mutable access, or a const reference to get a pointer to large data without copying.

rohanchanani committed:
cleaned up span choice and removed auto

rohanchanani commented on tests/deppart.cc:
```diff
@@ -161,6 +162,281 @@ int find_split(const std::vector<T> &cuts, T v)
+        return Event::NO_EVENT;
+  }
+
+  virtual int perform_dynamic_checks(void)
```
The test interface requires definitions for these virtual functions. I moved the checks to check partitioning and left perform_dynamic_checks as a no op.

rohanchanani commented on src/realm/deppart/sparsity_impl.cc:
```diff
@@ -457,6 +451,9 @@ namespace Realm {
   template <int N, typename T>
   class SparsityMapToRectAdapter {
   public:
+    SparsityMapToRectAdapter(const span<SparsityMapEntry<N, T>> &_entries)
```
I got rid of it. the original one is used by the cpu sort + merge pipeline, I don't actually need the span overload for the constructor.

rohanchanani commented on src/realm/deppart/partitions_gpu_impl.hpp:
```diff
@@ -0,0 +1,269 @@
+                         const Rect<N,T>& b) const {
+      Rect<N,T> r;
+      for(int d=0; d<N; d++){
+        r.lo[d] = a.lo[d] <  b.lo[d] ? a.lo[d] : b.lo[d];
```
nvcc errors when I use std::min and std::max in device code.

rohany replied:
What about just min/max? I think nvcc provides stubs for these.

rohany replied:
Did you undo this change?

rohanchanani committed:
cleanup

rohanchanani committed:
Switched to nvtx scoped range

rohanchanani committed:
Added deppart integration test

muraj committed:
kokkos: Explicitly set the kokkos language on the kokkos interop file (#270)

muraj committed:
Strip trailing whitespace from VERSION file

rohanchanani committed:
Made realm_malloc idiomatic

rohanchanani committed:
fixed cuda streams

rohanchanani committed:
moved pop bitmasks to execute

eddy16112 committed:
lazy creation of compiled lookup program (#273)

As discussed in slack, the `CompiledProgram` is only used when
MultiAffine accessor is created, so to avoid the exhaustion of
replicated heap, we can lazily create the `CompiledProgram` only when
`get_lookup_program` is called.

rohany commented on src/realm/deppart/byfield.cc:
```diff
@@ -322,21 +373,29 @@ namespace Realm {
+      for (size_t i = 0; i < colors.size(); i++) {
+        uop->add_sparsity_output(colors[i], subspaces[i]);
+      }
+      uop->dispatch(this, true);
```
I think so? I'm not sure what thread this code actually runs in... If it runs on a background worker then I think we're OK to stay on this thread, but otherwise we shouldn't.

rohany commented on .github/workflows/ci.yml:
```diff
@@ -29,7 +29,7 @@ concurrency:
 
 jobs:
   build:
```
What is this change?

rohanchanani replied:
I think from a commit I fetched in between.

muraj committed:
Fix up gasnet and apply the correct regex to the VERSION file's version (#274)

muraj committed:
Add min cuda arch to deal with kokkos compile issues (#281)

ebendler committed:
add ability to set realm.hpp namespace with a define. Fix realm_event… (#282)

…_wait call to use new declaration. Also some minor clang-format stuff.

muraj committed:
Fix gasnet linker flags (#288)

muraj committed:
Remove cudart and hip hijacks (#247)

muraj committed:
Add hash functions to id and some POD classes. (#286)

Closes #114.

rohanchanani committed:
reverted realm_malloc

muraj committed:
fix the coverage target to deal with changes in add_custom_command (#289)

muraj committed:
Fixup doxygen main page

rohanchanani commented on src/realm/sparsity.h:
```diff
@@ -246,8 +246,12 @@ namespace Realm {
+    span<SparsityMapEntry<N,T> > span_entries;
+    std::vector<Rect<N,T> > approx_rects;
+    span<Rect<N,T> > span_approx_rects;
+    RegionInstance entries_instance;
```
If I define the sparsity map impl destructor, it gets called twice - once I think from going out of scope, and the other from the wrapper calling map deleter (which just does delete impl). For now, I added a guard in the destructor to only destroy the instances the first time, but I could also define a pseudoconstructor that deletes the instance and add a call to that to map deleter.

rohanchanani committed:
ready for re-review

rohanchanani committed:
fixed destructor

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,268 @@
+    SparsityMapEntry<N, T>* d_parent_entries = reinterpret_cast<SparsityMapEntry<N,T>*>(AffineAccessor<char,1>(parent_entries_instance, 0).base);
+    CUDA_CHECK(cudaMemcpyAsync(d_parent_entries, parent_entries.data(), parent_entries.size() * sizeof(SparsityMapEntry<N,T>), cudaMemcpyHostToDevice, stream), stream);
+
+    //This is used for count + emit: first pass counts how many rectangles survive intersection, second pass uses the counter
```
Isnt this producing points, not rectangles?

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,268 @@
+        SparsityMapImpl<N, T> *impl = SparsityMapImpl<N, T>::lookup(it.second);
+        impl->gpu_finalize();
+      }
+      inst_offsets_instance.destroy();
```
I don't see some of these used in the `else` case until they are deleted later anyway. Can they just be unconditionally deleted?

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,268 @@
+        /* d_temp_storage */  nullptr,
+        /* temp_bytes */      rect_temp_bytes,
+        /* d_in */            d_volumes,
+        /* d_out */           d_prefix_rects + 1,   // shift by one so prefix[1]..prefix[n]
```
Why are you shifting by one then doing an inclusive scan? Isn't that the same as a normal exclusive scan?

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,268 @@
+        /* d_out */           d_prefix_rects + 1,   // shift by one so prefix[1]..prefix[n]
+        /* num_items */       num_valid_rects, stream);
+
+    RegionInstance temp_instance = this->realm_malloc(rect_temp_bytes, my_mem);
```
What is this second scan + temporary?

rohanchanani replied:
Cub calls need two runs, the first is to figure out how big your temporary should be and the second actually does the comptuation.

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,268 @@
+    RegionInstance colors_instance;
+
+    //Memcpying a boolean vector breaks things for some reason so we have this disgusting workaround
+    if constexpr(std::is_same_v<FT,bool>) {
```
Is making `FT` a bool even allowed?

rohany replied:
The reason for this though is that `vector<bool>` is specialized to an implementation that packs 8 elements into a single byte, rather than a byte per element.

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,268 @@
+    //This is where the work is actually done - each thread figures out which points to read, reads it, marks a PointDesc with its color, and writes it out
+    int num_blocks = (total_pts + flattened_threads - 1) / flattened_threads;
+    byfield_gpuPopulateBitmasksKernel<N,T,FT><<<num_blocks, flattened_threads, 0, stream>>>(d_accessors, d_valid_rects, d_prefix_rects, d_inst_prefix, d_colors, total_pts, colors.size(), num_valid_rects, field_data.size(), d_points);
+    KERNEL_CHECK(stream);
```
Do you need separate `KERNEL_CHECK` and `CUDA_CHECK` macros?

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,268 @@
+    temp_instance.destroy();
+    inst_counters_instance.destroy();
+
+    RegionInstance points_instance = this->realm_malloc(total_pts * sizeof(PointDesc<N,T>), my_mem);
```
Stylistically, i would declare this variable closer to where it is used for the first time.

rohany commented on src/realm/deppart/byfield_gpu_impl.hpp:
```diff
@@ -0,0 +1,268 @@
+    inst_offsets[field_data.size()] = inst_size;
+
+
+    RegionInstance inst_entries_instance = this->realm_malloc(inst_size * sizeof(SparsityMapEntry<N,T>), my_mem);
```
If I remember the presentation correctly, could this first group of work (intersecting source space rectangles etc) be extracted into a separate helper that the other ops will use too?

rohany commented on src/realm/deppart/byfield_gpu_kernels.hpp:
```diff
@@ -0,0 +1,110 @@
+  }
+  size_t low = 0, high = numInsts;
+  while (low < high) {
+      size_t mid = (low + high) >> 1;
```
look up and do the "no overflow" binary search. This can overflow if low and high are too big. It's unlikely to happen here because numInsts is small but just to use the "correct" algorithm.

rohany commented on src/realm/deppart/byfield_gpu_kernels.hpp:
```diff
@@ -0,0 +1,110 @@
+      else                                 high = mid;
+  }
+  size_t inst_idx = low;
+  uint32_t local = atomicAdd(&d_inst_counters[inst_idx], 1);
```
When `d_rects` is not null, do you need to do the atomic add? It doesn't seem that way in the code above.

rohanchanani replied:
Yes because I prefix sum the counters do figure out how big to make the output and to know where to write each output rectangle in the emit.

rohany commented on src/realm/deppart/byfield_gpu_kernels.hpp:
```diff
@@ -0,0 +1,110 @@
+
+  // Binary search to find which instance this rectangle belongs to
+  low = 0, high = num_insts;
+  while (low < high) {
```
This is binary search #3, probably worth it to put it into a helper function now for clarity.

rohany commented on src/realm/deppart/partitions_gpu_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+  #define NVTX_CAT(a,b)  a##b
+
+  #define NVTX_DEPPART(message) \
+  nvtxScopedRange NVTX_CAT(nvtx_, message)("cuda", #message, 0)
```
shouldn't the message be `deppart` or something?

rohany commented on src/realm/deppart/partitions_gpu_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+    T* d_keys_out = reinterpret_cast<T*>(aux_out);
+
+    PointDesc<N,T>* d_points_in = d_points;
+    PointDesc<N,T>* d_points_out = reinterpret_cast<PointDesc<N,T>*>(AffineAccessor<char,1>(pg_instance, 0).base);
```
Why are you doing reinterpret cases instead of making the affine accessor with the right type?

rohany commented on src/realm/deppart/partitions_gpu_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+    size_t* d_src_keys_out = reinterpret_cast<size_t*>(aux_out);
+
+    size_t t1=0, t2=0, t3=0;
+    cub::DeviceRadixSort::SortPairs(nullptr, t1, d_keys_in, d_keys_out, d_points_in, d_points_out, total_pts, 0, 8*sizeof(T), stream);
```
what are these 8*T sizes?

rohany commented on src/realm/deppart/partitions_gpu_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+
+
+    //Sort along each dimension from LSB to MSB (0 to N-1)
+    int threads_per_block = 256;
```
yep, pull this into a macro near the other macros.

rohany commented on src/realm/deppart/partitions_gpu_impl.hpp:
```diff
@@ -0,0 +1,343 @@
+
+  template<int N, typename T>
+  template<typename Container, typename IndexFn, typename MapFn>
+  void GPUMicroOp<N,T>::send_output(RectDesc<N, T>* d_rects, size_t total_rects, Memory my_mem, const Container& ctr, IndexFn getIndex, MapFn getMap)
```
Let's add this part of the pipeline in a future PR when we need it.

rohanchanani replied:
Byfield uses send_output - it's called at the end of complete_pipeline.

rohany commented on src/realm/deppart/sparsity_impl.cc:
```diff
@@ -907,6 +902,19 @@ namespace Realm {
+{
+     //We are responsible for our instances
+     if (this->from_gpu) {
+       if (get_runtime()->get_memory_impl(this->entries_instance))
```
What is a case where the memory no longer exists but the instance does?

rohany commented on src/realm/deppart/sparsity_impl.cc:
```diff
@@ -907,6 +902,19 @@ namespace Realm {
+         this->entries_instance.destroy();
+       if (get_runtime()->get_memory_impl(this->approx_instance))
+         this->approx_instance.destroy();
+       this->from_gpu = false;
```
Why do you need to set `from_gpu` to be false here? The object is being destructed. Also, Why do you need a `from_gpu` bit in the first place? Isn't that apparent from the existence of the isntances?

rohany commented on src/realm/deppart/sparsity_impl.cc:
```diff
@@ -1494,17 +1501,16 @@ namespace Realm {
-          it != this->entries.end(); it++) {
-        if(it->bitmap) {
+      for(size_t i = 0; i < this->get_entries().size(); i++) {
+        SparsityMapEntry<N, T> entry = this->get_entries()[i];
```
probably `const SparsityMapEntry<N, T>& entry` here.

rohany commented on src/realm/deppart/sparsity_impl.cc:
```diff
@@ -1830,6 +1839,117 @@ namespace Realm {
+
+  //Allows a GPU deppart client to set the entries directly with a host region instance
+  template<int N, typename T>
+    void SparsityMapImpl<N, T>::set_instance(RegionInstance _entries_instance, size_t size)
```
If we're always using `set_instance` from the gpu deppart ops, why do we need to copy/expose a separate `gpu_finalize` function?

rohany commented on src/realm/indexspace.inl:
```diff
@@ -591,41 +591,38 @@ namespace Realm {
-        }
+      for(size_t i = 0; i < entries.size(); i++) {
+        SparsityMapEntry<N, T> entry = entries[i];
+	if(!entry.bounds.contains(p)) continue;
```
odd formatting

rohanchanani committed:
made inst space helper

rohanchanani committed:
made parent space helper

rohanchanani committed:
ungeneralized input helper

rohanchanani committed:
generalized intersect input rects

rohanchanani committed:
builds with bvh path (untested)

rohanchanani committed:
moved prefix sum into a helper

rohanchanani committed:
moved accessors into zcpymem

rohanchanani committed:
cleaned up

rohanchanani committed:
mixed instances working

rohanchanani commented:
@rohany @lightsighter @artempriakhin Here's a writeup for the two proposals on how to handle dynamic allocations:

https://docs.google.com/presentation/d/1JQrFJlsi_BNijg2eTmTd3DxqYCBQYjiZsqNl0FP8k10/edit?usp=sharing

Proposal 1: If an allocation fails/is deferred, recycle all the sparsity map outputs and set their ids to 0. The user is then responsible for checking the sparsity IDs of their output index spaces once the partitioning op event resolves and handling failure.
Proposal 2: Break each operation into multiple calls, with allocations between the calls based on calculated sizes. The total # of required functions for a single partitioning op range from 3 to 7, although many of the functions can be used for multiple ops.

lightsighter commented:
Rather than respond to each of the comments above for each different proposal, I'm going to focus on this:

> > Proposal 5: (from @streichler) reserve a small amount of GPU memory for all such dependent partitioning operations, optimistically try to allocate buffers (e.g. in the style of Proposal 1) and then fall back to a more serialized execution using the pre-allocated small buffer as necessary. Adding my own idea to this one: the buffer could even be in zero-copy or UVM memory in order to avoid unnecessarily consuming device-side memory.

> I think this makes a lot of sense. One consideration is that the lifetime of GPU deppart operations will generally be a small fraction of the application's runtime, but they need a meaningful chunk of memory for their small burst - maybe this could include a way for the client to reclaim the memory when they know they aren't doing any partitioning?

If it is possible for the code to potentially work out of a smaller buffer as well as running faster with a larger buffer, then I think the interface I would like would be a two-phase model. In the first phase we do some kind of API call that estimates bounds on memory usage and returns two sizes: a minimal size needed to run the computation and an "optimal" size that ensures that deppart can get "good" performance (I feel like this value should be more a function of the GPU being targeted and not a property of the data being used for the deppart operation, but tell me if I am wrong about that assumption). If Realm is going to run on CPUs it will just return zero for both values (or maybe a small value, not sure if we use buffers on the CPU implementation). The client can then do an allocation at either value (or somewhere in-between if they're willing to search) and pass that in to the implementation. 

Tell me if that is possible and if not then we'll go back to exploring the other proposals.

> Legion is by far the biggest consumer of Realm, so I'll gladly defer to whatever combination of these proposals you think would be the most natural/intuitive for incorporating GPU-deppart into Legion while maintaining Realm's invariants.

Even though I care a lot about the Legion implementation, we've always designed Realm in a way so that it is Legion-agnostic. Realm should be usable by lots of different clients (I'm actually working on a different project built on Realm right now, it's just not public) so we try very hard to maintain that independence when designing APIs for Realm.

lightsighter commented:
> Say you want to do the computation with memory size K such that K < "optimal" size.

Can you say what your definition of the "optimal size" is? To be fair, let me give you my definition of "optimal size". In general, all of these dependent partitioning operations are going to be memory bandwidth limited. You can saturate the memory bandwidth of a GPU with many fewer SMs than most GPUs have (only really tiny GPUs are balanced and you'll mostly find them in laptops). Therefore, my definition of "optimal size" is however much memory I would need to allocate in order to have a big enough working set for the number of SMs required to saturate memory bandwidth. After that, everything else should just be executed in serial because even if I "parallelize" it with more blocks/threads, it's also going to be implicitly serialized anyway by the memory controller for the HBM memory. 

> Any accommodation I can think of to these challenges ends up serializing things to the point of defeating the purpose of gpu partitioning

If you serialize things only after having saturated memory bandwidth, then you're not going to lose any performance.

lightsighter commented:
To further elaborate, you can do all this with a single kernel launch with a small number of threadblocks and using a [cooperative kernel](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXECUTION.html#group__CUDART__EXECUTION_1g504b94170f83285c71031be6d5d15f73) and then the GPU version of the algorithm a bit like the CPU algorithm with the inner loop(s) parallelized across some number of threads/threadblocks with necessary synchronization inserted using [cooperative groups](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#cooperative-groups). You can then scale it up or down as necessary depending on the available memory bandwidth of the particular GPU you're running on. Bonus points if you can make it use temporal locality to leverage the L2 cache.

rohanchanani commented:
What I mean by “optimal” size is enough to keep all the data the partitioning operation needs on the device for the entire operation. 

> Say you want to do the computation with memory size K such that K < "optimal" size.

What I mean here is that in order to to use any data less than the “optimal” size, you have to tile the operation in some way. 

> If you serialize things only after having saturated memory bandwidth, then you're not going to lose any performance.

In the approach you’re describing, the tilling would be such that the total amount of work done stays the same, but the active working set at any given time is less. Thus, there’d only be performance degradation if you’re not saturating memory bandwidth, but no matter what the total amount of computation you’re doing would be the same (just N pieces of size 1 / N rather than 1 whole piece).  

I don’t think it’s possible to tile the operations in this way because of the last step, where all the points/rectangles are coalesced into a sparsity map, because the points/rectangles aren’t independent (they need to know if they duplicate/overlap any others). If you split the operation into N pieces, you’d have to do the following: coalesce the first 1/N points, keep that active somewhere, coalesce the next 1/N points, merge that into the working set (more expensive than the coalescing), coalesce the next 1/N points, merge that, etc. So rather than doing the same amount of work with a smaller working set, this adds a meaningful amount of computation that wouldn’t have originally been done with 1 big piece (all the merge operations). Each merge operation would have to do the slide 19 algorithm on the combination of the two spaces: https://docs.google.com/presentation/d/1Iwo0IwXBk14-E8i5i5kLN2B0KP7Cl-RwC0yDIN0_Fh8/edit?usp=sharing

> Any accommodation I can think of to these challenges ends up serializing things to the point of defeating the purpose of gpu partitioning, although I'm definitely open to if there's a better way to do things in less than the "optimal" size.

Here I don’t mean serialized in terms of less block/thread parallelism, but rather in terms of 1 bulk operation serialized into N tile operations, each of which requires a new, expensive integration step in between. And the nature of the computation goes from something that’s well-suited for the GPU (dumping a bulk operation onto the device and treating its elements as independent until the last possible moment) to something that’s generally better suited for the CPU (maintaining an active, disjoint state and iteratively merging pieces into it, which is also exactly what the CPU deppart path currently does). 

> But to know the memory used at stage N for chunk C, you need to have already computed each stage n < N.

I also still think this is a key challenge for any tilling approach with a fixed size buffer. Given a fixed size you have to stay within, I don’t think it’s possible to efficiently choose an input chunk size that's guaranteed to stay within the given size for the entire operation, i.e. to efficiently invert the pipeline. So when you choose a tile, there's a chance that tile will get to say the third stage and increase to bigger than your buffer, at which point the operation would have to fail.

If it's helpful, we could also hop on a call to discuss this.

rohany commented:
> Presuming you're keeping the rectangle data in shared memory, you can even have all the threads looking for conflicts by going through the [distributed shared memory](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#distributed-shared-memory) network across blocks so you won't incur any extra global memory traffic.

I don't think that it's appropriate to try and use DSMEM here, as only 2 "adjacent" CTAs can access each others' shared memory.

Taking a step back, I want to understand the root of your concerns here. A hybrid algorithm like the one you describe may be possible (@rohanchanani will be the arbiter of that), but what's the main reason that you are pushing for it? Is it mostly for avoiding the complexity of adding the multi-stage deppart API in Realm, or because you think that we should just doing better with the algorithmic components themselves? I am a bit worried about the complexity and cpu-side overheads (launching lots of kernels and copies) of doing a hybrid algorithm with a small tile size -- if it's just performance we can use green contexts or something to limit the number of SM's and allow other application work to proceed. If it is to avoid the multi-stage API in Realm, perhaps @rohanchanani can try to develop an algorithm for one of the deppart operations that does work in a tiled manner and see what the relative performance and code complexity tradeoffs are.

It also seems like if there is even a single deppart operation that we wouldn't be able to apply this to (like maybe the dynamic allocations for the BVH in preimage can't be made to fit in a fixed buffer) and we have to add a staged API then we might as well add it for all deppart operations.

lightsighter commented:
> what's the main reason that you are pushing for it?

The only way for Legion to implement an unbounded memory requirement operation is to effectively map an unbounded pool in that memory which blocks all downstream operations from allocation/deallocating in that memory. I don't want that to be the default for GPU-accelerated deppart operations in Legion. Any amount of complexity is worth it to avoid that as the effects ripple far up the software stack (beyond Legion). I could care less about the implementation complexity in Legion, but as soon as the consequences become visible to Legion users then I care A LOT and am willing to go to any ends to avoid it if possible.

> if it's just performance we can use green contexts or something to limit the number of SM's and allow other application work to proceed. If it is to avoid the multi-stage API in Realm, perhaps @rohanchanani can try to develop an algorithm for one of the deppart operations that does work in a tiled manner and see what the relative performance and code complexity tradeoffs are

It's neither. It's the fact that the memory requirements ripple up the software stack and are visible to mappers and impact unrelated operations which is something that is not currently the case for deppart operations.

> In particular, if we need tiles in the 10%+ capacity of the GPU to actually get reasonable performance out of this pathway then that already feels like too much to carve out into a fixed pool and we'd want to not have the application be forced to make this choice a-priori.

You should be able to do this with way less than 10% of memory. I bet having a buffer on the order of 10-100MB is enough.

> It also seems like if there is even a single deppart operation that we wouldn't be able to apply this to (like maybe the dynamic allocations for the BVH in preimage can't be made to fit in a fixed buffer) and we have to add a staged API then we might as well add it for all deppart operations.

Also strong disagree here. If we have to let consequences ripple up the software stack, then we should tightly bound the kinds of operations that it applies to.

rohanchanani commented:
I can definitely get started on a tiled approach for the point-based pipelines - also, preimage range dumps into the points pipeline, so of the non-setops only image range requires rects. But I also think that the back-end of tiling will end up looking pretty close to identical for both points and rects because we'll want to store the active set as rectangles.

lightsighter commented:
> I finished a tiled implementation of image that should also work for the other operations. 

Congrats! Very excited about this!

> I will be able to give a "minimal size" to complete the operation and an "optimal size"

Three questions about this:
1. Will the "optimal" size be guaranteed to do the whole operation in one "tile" or is it just a best guess?
2. Did you propose an API for querying this somewhere? I didn't immediately see it. Might be easier to spot it if you merged from `main` so there aren't so many outstanding diffs.
3. If the user picks a buffer size, can they just linearly interpolate how many tiles will be needed, or will we need to measure that at runtime and report it in a profiling response from the deppart operation?

> although it seems the optimal size will be based on what it takes to do the operation in one go rather than what it takes to saturate memory bandwidth as the cost of tilling is significant.

Two questions about this:
1. Can you elaborate on the sources of the failures? I missed the first half of the talk last week. I can see how we might do redundant work in deduplicating rectangles/points in constructing the final sparsity map, but I'm not clear on how things might fail in the middle.
4. Can you plot (or just report) the "effective memory bandwidth" that is achieved even with the wasted/redundant work? It would be good to confirm that we're actually hitting the memory bandwidth limit in all cases, even if the result is that some (a lot) of that bandwidth is going to waste. Maybe we discover we're actually compute bound (a pleasant surprise although doubtful) or that there is some other limiting factor (e.g. branching/memory divergence) which we might be able to improve in some way (e.g. sorting the way we do the work).

> I'm also working on a host fallback for when the output itself doesn't fit in the device buffer, but the code above is the primary consideration for feedback/design decisions.

Seems reasonable. This path just needs to be functional and not necessarily fast.
